ENH: testing: parameter to toggle comparing a scalar with an array?

Currently, the private xp test functions cannot compare a scaler with an array, even though we are ultimately using `np.testing`, we check for the equality of the total number of elements before:

https://github.com/data-apis/array-api-extra/blob/5045afa282b3956607265b1d6545898b23524278/src/array_api_extra/_lib/_testing.py#L84

To align more with numpy testing behaviour, we can use the strict parameter to check for this. If set to true, do the above check, if False check if it can broadcast. See the numpy example [here](https://numpy.org/devdocs/reference/generated/numpy.testing.assert_array_equal.html#:~:text=As%20mentioned%20in%20the%20Notes%20section%2C%20assert_array_equal%20has%20special%20handling%20for%20scalars.%20Here%20the%20test%20checks%20that%20each%20value%20in%20x%20is%203:). 

An even cleaner solution would be to remove the above check entirely and pass the `strict` parameter directly to the np.testing function call. The current numpy behaviour is different from what's suggested by these comments:

https://github.com/data-apis/array-api-extra/blob/5045afa282b3956607265b1d6545898b23524278/src/array_api_extra/_lib/_testing.py#L81

Either it refers to an older version of numpy testing, or I'm missing something.

Yes, we have some 0D cases in SciPy (not yet sure about sklearn):

https://github.com/scipy/scipy/blob/d4d823ac98ae301051617a5ed18579cae1f2f175/scipy/signal/tests/test_filter_design.py#L3085

https://github.com/scipy/scipy/blob/d4d823ac98ae301051617a5ed18579cae1f2f175/scipy/signal/tests/test_filter_design.py#L1196

https://github.com/scipy/scipy/blob/d4d823ac98ae301051617a5ed18579cae1f2f175/scipy/interpolate/tests/test_interpolate.py#L1778

This is the complete list:

```

FAILED scipy/interpolate/tests/test_bsplines.py::TestInterop::test_splint - AssertionError: sizes do not match: 6 != f1
FAILED scipy/interpolate/tests/test_bsplines.py::TestNdBSpline::test_2D_separable_2[numpy] - AssertionError: sizes do not match: 12 != f3
FAILED scipy/interpolate/tests/test_bsplines.py::TestNdBSpline::test_2D_separable_2[array_api_strict] - AssertionError: sizes do not match: 12 != f3
FAILED scipy/interpolate/tests/test_interpolate.py::TestInterp1D::test_bounds - AssertionError: sizes do not match: 4 != f1
FAILED scipy/interpolate/tests/test_interpolate.py::TestInterp1D::test_fill_value - AssertionError: sizes do not match: 6 != f1
FAILED scipy/interpolate/tests/test_interpolate.py::TestPPoly::test_roots_random[numpy] - AssertionError: sizes do not match: 2 != f1
FAILED scipy/signal/tests/test_filter_design.py::TestFreqz_sos::test_freqz_sos_design[numpy] - AssertionError: sizes do not match: 52 != f1
FAILED scipy/signal/tests/test_filter_design.py::TestFreqz_sos::test_freqz_sos_design[array_api_strict] - AssertionError: sizes do not match: 52 != f1
FAILED scipy/signal/tests/test_filter_design.py::TestButter::test_basic[numpy] - AssertionError: sizes do not match: 0 != f1
FAILED scipy/signal/tests/test_filter_design.py::TestButter::test_basic[array_api_strict] - AssertionError: sizes do not match: 0 != f1

```

10 failures. It's actually not too many. One workaround (in scipy tests) would be to use `xp.full_like` and broadcast the 0d array.

I have some insights on `scikit-learn`'s testing mechanism.
-  It relies completely on np.testing functions. All the inputs are converted to a numpy array using ` actual, desired = np.asanyarray(actual), np.asanyarray(desired)`. Hence, the more the xp test function adheres to numpy behaviour, the easier the transition. 
- When I replace the `np_assert_allclose` in this function with `xp_assert_close` from `array_api_extra`, we get close to `238` failures, most of them are due to: `AssertionError: sizes do not match`
 - The 0d input that we talked about earlier, sklearn uses it quite often. See sklearn's `assert_allclose` function [here](https://github.com/scikit-learn/scikit-learn/blob/e67ae2f1412a4c67ce4c24061ce44af2af3d5bcb/sklearn/utils/_testing.py#L174). Here are a few places where it's used :

https://github.com/scikit-learn/scikit-learn/blob/e67ae2f1412a4c67ce4c24061ce44af2af3d5bcb/sklearn/tree/tests/test_monotonic_tree.py#L99

https://github.com/scikit-learn/scikit-learn/blob/e67ae2f1412a4c67ce4c24061ce44af2af3d5bcb/sklearn/semi_supervised/tests/test_label_propagation.py#L94

List of all files with similar comparisons (all thanks to AI for extracting these from the logs): 
```
sklearn/cluster/tests/test_affinity_propagation.py
sklearn/decomposition/tests/test_truncated_svd.py
sklearn/ensemble/tests/test_weight_boosting.py
sklearn/gaussian_process/tests/test_gpr.py
sklearn/impute/tests/test_impute.py
sklearn/linear_model/tests/test_base.py
sklearn/linear_model/tests/test_coordinate_descent.py
sklearn/metrics/tests/test_classification.py
sklearn/metrics/tests/test_pairwise.py
sklearn/metrics/tests/test_ranking.py
sklearn/model_selection/tests/test_search.py
sklearn/model_selection/tests/test_validation.py
sklearn/preprocessing/tests/test_data.py
sklearn/semi_supervised/tests/test_label_propagation.py
sklearn/tests/test_common.py
sklearn/tests/test_pipeline.py
sklearn/tree/tests/test_monotonic_tree.py
sklearn/utils/tests/test_estimator_checks.py
```

- I just added this line to avoid the 0d (else case from the above discussion) comparison:

```py

    if desired.shape != actual.shape and not check_shape and desired.ndim == 0:
        desired = np.broadcast_to(desired, actual.shape)

```
It brought down the total errors to just 2! 

- We have two options: either fix the `sklearn` function (adding the above line) or remove the else block in `array_api_extra`, both of which are simple fixes. 
- I feel that our testing functions are good enough to be able to get adopted 
- In case of sklearn, it's pretty easy to switch to `xp_assert` functions since they have heavy reliance on numpy style testing behaviour, in case of scipy, they have their own set of strict testing of inputs, but can still be adapted, as `array_api_strict` also has most of those tests.

@lucascolley Would love your thoughts here.

--- a/src/array_api_extra/_lib/_testing.py
+++ b/src/array_api_extra/_lib/_testing.py
@@ -37,7 +37,7 @@ def _check_ns_shape_dtype(
     check_dtype: bool,
     check_shape: bool,
     check_scalar: bool,
-) -> ModuleType:  # numpydoc ignore=RT03
+) -> tuple[Array, Array, ModuleType]:  # numpydoc ignore=RT03
     """
     Assert that namespace, shape and dtype of the two arrays match.
 
@@ -55,7 +55,7 @@ def _check_ns_shape_dtype(
 
     Returns
     -------
-    Arrays namespace.
+    Actual array, desired array, and their namespace.
     """
     actual_xp = array_namespace(actual)  # Raises on scalars and lists
     desired_xp = array_namespace(desired)
@@ -63,6 +63,16 @@ def _check_ns_shape_dtype(
     msg = f"namespaces do not match: {actual_xp} != f{desired_xp}"
     assert actual_xp == desired_xp, msg
 
+    if is_numpy_namespace(actual_xp) and check_scalar:
+        # only NumPy distinguishes between scalars and arrays; we do if check_scalar.
+        _msg = (
+            "array-ness does not match:\n Actual: "
+            f"{type(actual)}\n Desired: {type(desired)}"
+        )
+        assert np.isscalar(actual) == np.isscalar(desired), _msg
+
+    actual = desired_xp.asarray(actual)
+    desired = desired_xp.asarray(desired)
     # Dask uses nan instead of None for unknown shapes
     actual_shape = cast(tuple[float, ...], actual.shape)
     desired_shape = cast(tuple[float, ...], desired.shape)
@@ -90,16 +100,8 @@ def _check_ns_shape_dtype(
     if check_dtype:
         msg = f"dtypes do not match: {actual.dtype} != {desired.dtype}"
         assert actual.dtype == desired.dtype, msg
-
-    if is_numpy_namespace(actual_xp) and check_scalar:
-        # only NumPy distinguishes between scalars and arrays; we do if check_scalar.
-        _msg = (
-            "array-ness does not match:\n Actual: "
-            f"{type(actual)}\n Desired: {type(desired)}"
-        )
-        assert np.isscalar(actual) == np.isscalar(desired), _msg
-
-    return desired_xp
+    desired = desired_xp.broadcast_to(desired, actual.shape)
+    return actual, desired, desired_xp
 
 
 def _is_materializable(x: Array) -> bool:
@@ -169,7 +171,9 @@ def xp_assert_equal(
     xp_assert_close : Similar function for inexact equality checks.
     numpy.testing.assert_array_equal : Similar function for NumPy arrays.
     """
-    xp = _check_ns_shape_dtype(actual, desired, check_dtype, check_shape, check_scalar)
+    actual, desired, xp = _check_ns_shape_dtype(
+        actual, desired, check_dtype, check_shape, check_scalar
+    )
     if not _is_materializable(actual):
         return
     actual_np = as_numpy_array(actual, xp=xp)
@@ -211,7 +215,7 @@ def xp_assert_less(
     xp_assert_close : Similar function for inexact equality checks.
     numpy.testing.assert_array_equal : Similar function for NumPy arrays.
     """
-    xp = _check_ns_shape_dtype(x, y, check_dtype, check_shape, check_scalar)
+    x, y, xp = _check_ns_shape_dtype(x, y, check_dtype, check_shape, check_scalar)
     if not _is_materializable(x):
         return
     x_np = as_numpy_array(x, xp=xp)
@@ -267,7 +271,9 @@ def xp_assert_close(
     -----
     The default `atol` and `rtol` differ from `xp.all(xpx.isclose(a, b))`.
     """
-    xp = _check_ns_shape_dtype(actual, desired, check_dtype, check_shape, check_scalar)
+    actual, desired, xp = _check_ns_shape_dtype(
+        actual, desired, check_dtype, check_shape, check_scalar
+    )
     if not _is_materializable(actual):
         return
 

--- a/src/array_api_extra/_lib/_testing.py
+++ b/src/array_api_extra/_lib/_testing.py
@@ -78,9 +78,11 @@ def _check_ns_shape_dtype(
     assert None not in desired_shape
     if is_dask_namespace(desired_xp):
         if any(math.isnan(i) for i in actual_shape):
-            actual_shape = actual.compute().shape  # type: ignore[attr-defined]  # pyright: ignore[reportAttributeAccessIssue]
+            actual.compute_chunk_sizes()  # type: ignore[attr-defined]  # pyright: ignore[reportAttributeAccessIssue]
+            actual_shape = actual.shape
         if any(math.isnan(i) for i in desired_shape):
-            desired_shape = desired.compute().shape  # type: ignore[attr-defined]  # pyright: ignore[reportAttributeAccessIssue]
+            desired.compute_chunk_sizes()  # type: ignore[attr-defined]  # pyright: ignore[reportAttributeAccessIssue]
+            desired_shape = desired.shape
 
     if check_shape:
         msg = f"shapes do not match: {actual_shape} != f{desired_shape}"

--- a/src/array_api_extra/_lib/_testing.py
+++ b/src/array_api_extra/_lib/_testing.py
@@ -98,7 +98,7 @@ def _check_ns_shape_dtype(
     if check_dtype:
         msg = f"dtypes do not match: {actual.dtype} != {desired.dtype}"
         assert actual.dtype == desired.dtype, msg
-    desired = desired_xp.broadcast_to(desired, actual.shape)
+    desired = desired_xp.broadcast_to(desired, actual_shape)
     return actual, desired, desired_xp
 
 

--- a/src/array_api_extra/_lib/_testing.py
+++ b/src/array_api_extra/_lib/_testing.py
@@ -57,7 +57,7 @@ def _check_ns_shape_dtype(
     -------
     Actual array, desired array, and their namespace.
     """
-    actual_xp = array_namespace(actual)  # Raises on scalars and lists
+    actual_xp = array_namespace(actual)  # Raises on Python scalars and lists
     desired_xp = array_namespace(desired)
 
     msg = f"namespaces do not match: {actual_xp} != f{desired_xp}"

--- a/src/array_api_extra/_lib/_testing.py
+++ b/src/array_api_extra/_lib/_testing.py
@@ -76,13 +76,14 @@ def _check_ns_shape_dtype(
     desired_shape = cast(tuple[float, ...], desired.shape)
     assert None not in actual_shape  # Requires explicit support
     assert None not in desired_shape
+
     if is_dask_namespace(desired_xp):
         if any(math.isnan(i) for i in actual_shape):
             actual.compute_chunk_sizes()  # type: ignore[attr-defined]  # pyright: ignore[reportAttributeAccessIssue]
-            actual_shape = actual.shape
+            actual_shape = cast(tuple[float, ...], actual.shape)
         if any(math.isnan(i) for i in desired_shape):
             desired.compute_chunk_sizes()  # type: ignore[attr-defined]  # pyright: ignore[reportAttributeAccessIssue]
-            desired_shape = desired.shape
+            desired_shape = cast(tuple[float, ...], desired.shape)
 
     if check_shape:
         msg = f"shapes do not match: {actual_shape} != f{desired_shape}"
@@ -92,8 +93,8 @@ def _check_ns_shape_dtype(
         # np.testing.assert_array_equal etc even when strict=False, but not for
         # non-materializable arrays.
         # This check excludes 0d arrays as they are special-cased in NumPy.
-        actual_size = math.prod(actual_shape)  # pyright: ignore[reportUnknownArgumentType]
-        desired_size = math.prod(desired_shape)  # pyright: ignore[reportUnknownArgumentType]
+        actual_size = math.prod(actual_shape)
+        desired_size = math.prod(desired_shape)
         msg = f"sizes do not match: {actual_size} != f{desired_size}"
         assert actual_size == desired_size, msg
 

--- a/src/array_api_extra/_lib/_testing.py
+++ b/src/array_api_extra/_lib/_testing.py
@@ -71,8 +71,6 @@ def _check_ns_shape_dtype(
         )
         assert np.isscalar(actual) == np.isscalar(desired), _msg
 
-    actual = desired_xp.asarray(actual)
-    desired = desired_xp.asarray(desired)
     # Dask uses nan instead of None for unknown shapes
     actual_shape = cast(tuple[float, ...], actual.shape)
     desired_shape = cast(tuple[float, ...], desired.shape)

